subsys/bluetooth/services/ots/ots_dir_list.c

```diff
@@ -79,106 +123,132 @@ static void dir_list_object_encode(struct bt_gatt_ots_object *obj,
 	sys_put_le16(len, start);
 }
 
-void bt_ots_dir_list_obj_add(struct bt_ots_dir_list *dir_list, void *obj_manager,
-			     struct bt_gatt_ots_object *cur_obj, struct bt_gatt_ots_object *obj)
+static void bt_ots_dir_list_reset_anchor(struct bt_ots_dir_list *dir_list, void *obj_manager)
+{
+	dir_list->anchor_offset = 0;
+	bt_gatt_ots_obj_manager_first_obj_get(obj_manager, &dir_list->anchor_object);
+}
+
+static int bt_ots_dir_list_search_forward(struct bt_ots_dir_list *dir_list, void *obj_manager,
+					  off_t offset)
 {
-	__ASSERT(dir_list->dir_list_obj != obj,
-		 "Cannot add Directory Listing Object");
+	int err;
+	char id_str[BT_OTS_OBJ_ID_STR_LEN];
+	struct bt_gatt_ots_object *obj = dir_list->anchor_object;
+	size_t rec_len = dir_list_object_record_size(obj);
 
-	__ASSERT(bt_gatt_ots_obj_manager_obj_contains(obj_manager, obj),
-		 "Object not part of OTS instance");
+	bt_ots_obj_id_to_str(obj->id, id_str, sizeof(id_str));
+	LOG_DBG("Searching forward for offset 0x%lx starting at 0x%lx with object ID %s",
+		(long)offset, (long)dir_list->anchor_offset, log_strdup(id_str));
 
-	if (dir_list->dir_list_obj != cur_obj) {
-		/* We only need to update the object directory listing if it is currently selected,
-		 * as we otherwise only create it when it is selected.
-		 */
-		return;
-	}
+	while (dir_list->anchor_offset + rec_len <= offset) {
 
-	dir_list_object_encode(obj, &dir_list->net_buf);
+		err = bt_gatt_ots_obj_manager_next_obj_get(obj_manager, obj, &obj);
+		if (err) {
+			return -EINVAL;
+		}
 
-	/* Update Directory Listing Object metadata size */
-	dir_list->dir_list_obj->metadata.size.cur = dir_list->net_buf.len;
+		dir_list->anchor_offset += rec_len;
+		dir_list->anchor_object = obj;
+
+		rec_len = dir_list_object_record_size(obj);
+	}
+
+	return 0;
 }
 
-void bt_ots_dir_list_obj_remove(struct bt_ots_dir_list *dir_list, void *obj_manager,
-				struct bt_gatt_ots_object *cur_obj, struct bt_gatt_ots_object *obj)
+static int bt_ots_dir_list_search_backward(struct bt_ots_dir_list *dir_list, void *obj_manager,
+					   off_t offset)
 {
-	uint16_t offset = 0;
+	int err;
+	char id_str[BT_OTS_OBJ_ID_STR_LEN];
+	struct bt_gatt_ots_object *obj = dir_list->anchor_object;
 
-	__ASSERT(dir_list->dir_list_obj != obj,
-		 "Cannot remove Directory Listing Object");
+	bt_ots_obj_id_to_str(obj->id, id_str, sizeof(id_str));
+	LOG_DBG("Searching backward for offset 0x%lx starting at 0x%lx with object ID %s",
+		(long)offset, (long)dir_list->anchor_offset, log_strdup(id_str));
```

Comment:
Hmmmm ok, yeah I am always confused by `off_t` as well :( but I will change all the `off_t`s to be `%ld`, really it didn't matter to me as long as I could easily compare them while debugging. If the number has become negative, it will be obvious while debugging.

Reply:
Ok, but I still have to cast it as `long` since depending on the platform the size is indeterministic and unlike `size_t` there is no special format specifier :(

Reply:
Indeed. `off_t` is not part of the C standard, but rather the POSIX standard as far as I can tell. Not sure if it really is a type that perhaps should be avoided. @jhedberg Any input on the use of `off_t`?